expression/generator/compare_vec.go

```diff
@@ -0,0 +1,407 @@
+// limitations under the License.
+
+// +build ignore
+
+package main
+
+import (
+	"bytes"
+	"flag"
+	"go/format"
+	"io/ioutil"
+	"log"
+	"path/filepath"
+	"text/template"
+)
+
+const header = `// Copyright 2019 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Code generated by go generate in expression/generator; DO NOT EDIT.
+
+package expression
+`
+
+const newLine = "\n"
+
+const builtinCompareImports = `import (
+	"math"
+
+	"github.com/pingcap/parser/mysql"
+	"github.com/pingcap/tidb/types"
+	"github.com/pingcap/tidb/types/json"
+	"github.com/pingcap/tidb/util/chunk"
+)
+`
+
+var builtinCompareVecTpl = template.Must(template.New("").Parse(`
+func (b *builtin{{ .compare.CompareName }}{{ .type.TypeName }}Sig) vecEvalInt(input *chunk.Chunk, result *chunk.Column) error {
+	n := input.NumRows()
+	buf0, err := b.bufAllocator.get(types.ET{{ .type.ETName }}, n)
+	if err != nil {
+		return err
+	}
+	defer b.bufAllocator.put(buf0)
+	if err := b.args[0].VecEval{{ .type.TypeName }}(b.ctx, input, buf0); err != nil {
+		return err
+	}
+	buf1, err := b.bufAllocator.get(types.ET{{ .type.ETName }}, n)
+	if err != nil {
+		return err
+	}
+	defer b.bufAllocator.put(buf1)
+	if err := b.args[1].VecEval{{ .type.TypeName }}(b.ctx, input, buf1); err != nil {
+		return err
+	}
+
+{{ if .type.Fixed }}
+	arg0 := buf0.{{ .type.TypeNameInColumn }}s()
+	arg1 := buf1.{{ .type.TypeNameInColumn }}s(){{ end }}
+	result.ResizeInt64(n, false)
+	i64s := result.Int64s()
+{{ if eq .type.ETName "Int" }}
+	isUnsigned0 := mysql.HasUnsignedFlag(b.args[0].GetType().Flag)
+	isUnsigned1 := mysql.HasUnsignedFlag(b.args[1].GetType().Flag)
+{{ end }}
+	for i := 0; i < n; i++ {
+		if buf0.IsNull(i) || buf1.IsNull(i) {
+			result.SetNull(i, true)
+			continue
+		}
+{{ if eq .type.ETName "Int" }}
+		var val int64
+		switch {
+		case isUnsigned0 && isUnsigned1:
+			val = int64(types.CompareUint64(uint64(arg0[i]), uint64(arg1[i])))
+		case isUnsigned0 && !isUnsigned1:
+			if arg1[i] < 0 || uint64(arg0[i]) > math.MaxInt64 {
+				val = int64(1)
+			} else {
+				val = int64(types.CompareInt64(arg0[i], arg1[i]))
+			}
+		case !isUnsigned0 && isUnsigned1:
+			if arg0[i] < 0 || uint64(arg1[i]) > math.MaxInt64 {
+				val = int64(-1)
+			} else {
+				val = int64(types.CompareInt64(arg0[i], arg1[i]))
+			}
+		case !isUnsigned0 && !isUnsigned1:
+			val = int64(types.CompareInt64(arg0[i], arg1[i]))
+		}
+{{ else if eq .type.ETName "Json" }}
+		val := int64(json.CompareBinary(buf0.GetJSON(i), buf1.GetJSON(i)))
```

Comment:
Why should we cast to int64 here?

Reply:
I'm confused too.
Why do you use `int64` instead of `int`?